ir/src/ir2/graph/nodes.rs

```diff
@@ -0,0 +1,236 @@
+        }
+    }
+    fn new_i32(&mut self, data: i32) -> Link<NodeTypes> {
+        match self.borrow_mut().deref_mut() {
+            NodeTypes::Graph(graph) => graph.new_i32(data),
+            NodeTypes::Node(node) => node.new_i32(data),
+            NodeTypes::I32(_) => panic!("new_i32() called on leaf"),
+            NodeTypes::Add(add) => add.new_i32(data),
+            NodeTypes::Function(function) => function.body().new_i32(data),
+        }
+    }
+    fn new_add(&mut self) -> Link<NodeTypes> {
+        match self.borrow_mut().deref_mut() {
+            NodeTypes::Graph(graph) => graph.new_add(),
+            NodeTypes::Node(node) => node.new_add(),
+            NodeTypes::I32(_) => panic!("new_add() called on leaf"),
+            NodeTypes::Add(add) => add.new_add(),
+            NodeTypes::Function(function) => function.new_add(),
+        }
+    }
+    fn new_body(&mut self) -> Link<NodeTypes> {
+        match self.borrow_mut().deref_mut() {
+            NodeTypes::Graph(graph) => graph.new_body(),
+            NodeTypes::Node(_) => panic!("new_body() called on body"),
+            NodeTypes::I32(_) => panic!("new_body() called on leaf"),
+            NodeTypes::Add(_) => panic!("new_body() called on add"),
+            NodeTypes::Function(_) => panic!("new_body() called on function"),
+        }
+    }
+}
+
+impl Child for Link<NodeTypes> {
+    fn get_parent(&self) -> BackLink<NodeTypes> {
+        match self.borrow().deref() {
+            NodeTypes::Graph(_) => panic!("get_parent called on graph"),
+            NodeTypes::Node(node) => node.get_parent(),
+            NodeTypes::I32(leaf) => leaf.get_parent(),
+            NodeTypes::Add(add) => add.get_parent(),
+            NodeTypes::Function(function) => function.get_parent(),
+        }
+    }
+    fn swap_parent(&mut self, parent: Link<NodeTypes>) {
+        println!("Link<NodeTypes>.set_parent({:?}, {:?})", self, parent);
+        match self.borrow_mut().deref_mut() {
+            NodeTypes::Node(ref mut node) => node.swap_parent(parent),
+            NodeTypes::I32(ref mut leaf) => leaf.swap_parent(parent),
+            NodeTypes::Add(ref mut add) => add.swap_parent(parent),
+            NodeTypes::Function(ref mut function) => function.swap_parent(parent),
+            _ => panic!("set_parent called on non-node"),
+        }
+    }
+}
+
+#[derive(Clone, Eq, PartialEq)]
+pub struct Function {
+    node: Node,
+}
+
+impl Function {
+    pub fn new(args: Link<NodeTypes>, ret: Link<NodeTypes>, body: Link<NodeTypes>) -> Self {
+        Self {
+            node: Node::new(BackLink::none(), Link::new(vec![args, ret, body])),
+        }
+    }
+    pub fn args(&self) -> Link<NodeTypes> {
+        self.node.get_children().borrow().deref()[0].clone()
+    }
+    pub fn ret(&self) -> Link<NodeTypes> {
+        self.node.get_children().borrow().deref()[1].clone()
+    }
+    pub fn body(&self) -> Link<NodeTypes> {
+        self.node.get_children().borrow().deref()[2].clone()
```

Comment:
Keep in mind that index access can panic.
I don't think adding error handling is necessary, but maybe we can ensure that Function can only be created through Function::new() and without direct Object instantiation. Here, the field `node: Node` is private so it's fine, but on Add below it's not for example.

So nothing to change in this case, but just commenting to share :)